### 115."throws" declarations should not be superfluous


“throws”声明不应该抛出多个异常

An exception in a`throws`declaration in Java is superfluous if it is:

-   listed multiple times
-   a subclass of another listed exception
-   a`RuntimeException`, or one of its descendants
-   completely unnecessary because the declared exception type cannot actually be thrown


这些是Java中的throws声明中的异常是多余的：

- 多次列出
- 另一个列出的异常的子类
- 一个RuntimeException, 或者子类之一
- 完全没有必要，因为实际上无法引发已声明的异常类型


**Noncompliant Code Example**

**错误的代码示例**

```
void foo() throws MyException, MyException {}  // Noncompliant; should be listed once
void bar() throws Throwable, Exception {}  // Noncompliant; Exception is a subclass of Throwable
void baz() throws RuntimeException {}  // Noncompliant; RuntimeException can always be thrown
```

**Compliant Solution**

**正确的代码示例**


```
void foo() throws MyException {}
void bar() throws Throwable {}
void baz() {}
```

**Exceptions**

**例外**

The rule will not raise any issue for exceptions that cannot be thrown from the method body:

-   in overriding and implementation methods
-   in interface`default`methods
-   in non-private methods that only`throw`, have empty bodies, or a single return statement .
-   in overridable methods (non-final, or not member of a final class, non-static, non-private), if the exception is documented with a proper javadoc.


对于无法从方法主体引发的异常，该规则不会引起任何问题：

* 覆盖和实现方法
* 接口default方法中
* 非私有方法的throw，具有空主体或单个return语句。
* 如果可以使用适当的javadoc记录异常，则使用可重写的方法（非最终方法，或者不是最终类的成员，非静态，非私有）。


```
class A extends B {
  @Override
  void doSomething() throws IOException {
    compute(a);
  }

  public void foo() throws IOException {}

  protected void bar() throws IOException {
    throw new UnsupportedOperationException("This method should be implemented in subclasses");
  }

  Object foobar(String s) throws IOException {
    return null;
  }

  /**
   * @throws IOException Overriding classes may throw this exception if they print values into a file
   */
  protected void print() throws IOException { // no issue, method is overridable and the exception has proper javadoc
    System.out.println("foo");
  }
}
```


